idx/memory/tag_query.go

```diff
@@ -0,0 +1,408 @@
+
+import (
+	"errors"
+	"math"
+	"regexp"
+	"strings"
+
+	"github.com/grafana/metrictank/idx"
+)
+
+var (
+	errInvalidQuery = errors.New("invalid query")
+)
+
+const (
+	PARSING_ERROR = iota
+	EQUAL
+	NOT_EQUAL
+	MATCH
+	NOT_MATCH
+)
+
+type kv struct {
+	key   string
+	value string
+}
+
+type expression struct {
+	kv
+	operator int
+}
+
+type TagQuery struct {
+	from     int64
+	equal    []kv
+	match    []kv
+	notEqual []kv
+	notMatch []kv
+}
+
+// returns expression,
+// in case of error the operator will be PARSING_ERROR
+func parseExpression(expr string) expression {
+	var key []byte
+	var pos int
+	regex, not := false, false
+
+	// get key
+	for ; pos < len(expr); pos++ {
+		// ! || =
+		if expr[pos] == 33 || expr[pos] == 61 {
+			// key must not be empty
+			if len(key) == 0 {
+				return expression{operator: PARSING_ERROR}
+			}
+			break
+		}
+
+		// a-z A-Z 0-9
+		if !(expr[pos] >= 97 && expr[pos] <= 122) && !(expr[pos] >= 65 && expr[pos] <= 90) && !(expr[pos] >= 48 && expr[pos] <= 57) {
+			return expression{operator: PARSING_ERROR}
+		}
+
+		key = append(key, expr[pos])
+	}
+
+	// if !
+	if expr[pos] == 33 {
+		not = true
+		pos++
+	}
+
+	// expecting a =
+	if expr[pos] != 61 {
+		return expression{operator: PARSING_ERROR}
+	}
+	pos++
+
+	// if ~
+	if len(expr) > pos && expr[pos] == 126 {
+		regex = true
+		pos++
+	}
+
+	var value []byte
+	for ; pos < len(expr); pos++ {
+		// enforce a-z A-Z 0-9 -_ if query is not regex
+		if !regex && !(expr[pos] >= 97 && expr[pos] <= 122) && !(expr[pos] >= 65 && expr[pos] <= 90) && !(expr[pos] >= 48 && expr[pos] <= 57) && expr[pos] != 45 && expr[pos] != 95 {
+			return expression{operator: PARSING_ERROR}
+		}
+
+		value = append(value, expr[pos])
+	}
+
+	if not {
+		if regex {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: NOT_MATCH}
+		} else {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: NOT_EQUAL}
+		}
+	} else {
+		if regex {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: MATCH}
+		} else {
+			return expression{kv: kv{key: string(key), value: string(value)}, operator: EQUAL}
+		}
+	}
+}
+
+func NewTagQuery(expressions []string, from int64) (TagQuery, error) {
+	q := TagQuery{from: from}
+
+	if len(expressions) == 0 {
+		return q, errInvalidQuery
+	}
+
+	for _, expr := range expressions {
+		e := parseExpression(expr)
+		if e.operator == PARSING_ERROR {
+			return q, errInvalidQuery
+		}
+
+		// special case of empty value
+		if len(e.value) == 0 {
+			if e.operator == EQUAL || e.operator == MATCH {
+				q.notMatch = append(q.notMatch, kv{
+					key:   e.key,
+					value: "^.+",
+				})
+			} else {
+				q.match = append(q.match, kv{
+					key:   e.key,
+					value: "^.+",
+				})
+			}
+		} else {
+			// always anchor all regular expressions at the beginning
+			if (e.operator == MATCH || e.operator == NOT_MATCH) && len(e.value) > 0 && e.value[0] != byte('^') {
+				e.value = "^" + e.value
+			}
```

Comment:
This could surprise a user. Is this out of a concern for performance?

Reply:
It's the way that prometheus and graphite both work